teachertool/src/services/loggingService.ts

```diff
@@ -1,24 +1,45 @@
-const formatMessageForConsole = (message: string) => {
+const timestamp = () => {
     const time = new Date();
-    return `[${time.getHours()}:${time.getMinutes()}:${time.getSeconds()}] ${message}`;
-}
+    return `[${time.getHours()}:${time.getMinutes()}:${time.getSeconds()}]`;
+};
 
 const formatName = (name: string) => {
     return name.toLowerCase().replace(/ /g, "_");
-}
+};
 
-export const logError = (name: string, details: string) => {
-    pxt.tickEvent("teachertool.error", { name: formatName(name), message: details });
-    console.error(formatMessageForConsole(`${name}: ${details}`));
-}
+export const logError = (
+    errorCode: string,
+    message?: any,
+    data: pxt.Map<string | number> = {}
+) => {
+    errorCode = formatName(errorCode);
+    let dataObj = { ...data };
+    if (message) {
+        if (typeof message === "object") {
+            dataObj = { ...dataObj, ...message };
+            // Look for non-enumerable properties found on Error objects
+            ["message", "stack", "name"].forEach(key => {
```

Comment:
@thsparks this will log the callstack if `message` is of type `Error`.

Reply:
I don't know why these fields are non-enumerable. Makes no sense to me.

Reply:
hmm, would it work if you did an `if message instanceof Error` check?

Reply:
Nope, they'd still be non-enumerable and excluded from the spread operator.